fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/NewQueryBuilder.java

```diff
@@ -0,0 +1,529 @@
+        log.entering(CLASSNAME, METHODNAME,
+                new Object[] { resourceType.getSimpleName(), searchContext.getSearchParameters() });
+
+        final SearchQuery domainModel;
+
+        if (searchContext.hasSortParameters()) {
+            // Special variant of the query which will sort based on the given sort params
+            // and return a list of resource-ids which are then used to fetch the actual data
+            // (matching the old query builder design...for now).
+            SearchSortQuery sortQuery = new SearchSortQuery(resourceType.getSimpleName());
+            for (SortParameter sp: searchContext.getSortParameters()) {
+                sortQuery.add(new DomainSortParameter(sp));
+            }
+            domainModel = sortQuery;
+        } else {
+            domainModel = new SearchDataQuery(resourceType.getSimpleName());
+        }
+        buildModelCommon(domainModel, resourceType, searchContext);
+        Select result = renderQuery(domainModel, searchContext);
+
+        log.exiting(CLASSNAME, METHODNAME);
+        return result;
+    }
+
+    /**
+     * Builds a query that returns included resources.
+     *
+     * @param resourceType  - the type of resource being searched for.
+     * @param searchContext - the search context containing the search parameters.
+     * @param inclusionParm - the inclusion parameter for which the query is being built.
+     * @param ids           - the list of logical resource IDs the query will run against.
+     * @param inclusionType - either INCLUDE or REVINCLUDE.
+     * @return Select the query to fetch the matching list of included resources
+     * @throws Exception
+     */
+    public Select buildIncludeQuery(Class<?> resourceType, FHIRSearchContext searchContext,
+            InclusionParameter inclusionParm, List<Long> logicalResourceIds, String inclusionType) throws Exception {
+        final String METHODNAME = "buildIncludeQuery";
+        log.entering(CLASSNAME, METHODNAME);
+
+        // Build the special "include" query to fetch additional resources
+        // that need to be included with the main search result
+        final String includeResourceType;
+        if (SearchConstants.INCLUDE.equals(inclusionType)) {
+            log.fine("Building _include query");
+            includeResourceType = inclusionParm.getSearchParameterTargetType();
+        } else {
+            log.fine("Building _revinclude query");
+            includeResourceType = inclusionParm.getJoinResourceType();
+        }
+
+        // Start building a query model to fetch resources of the type we want to include
+        final SearchQuery domainModel;
+        if (SearchConstants.INCLUDE.equals(inclusionType)) {
+            // _include needs a different base in order to support versioned references
+            domainModel = new SearchIncludeQuery(includeResourceType);
+        } else {
+            domainModel = new SearchIncludeQuery(includeResourceType);
+        }
```

Comment:
IF and ELSE legs are the same

Reply:
Ah, the wonders of code evolution. Will fix.